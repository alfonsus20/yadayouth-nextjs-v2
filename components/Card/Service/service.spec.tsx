import { fireEvent, render, screen } from "@testing-library/react";
import mockRouter from "next-router-mock";
import singletonRouter from 'next/router';
import { ServiceCard } from "..";
import MockedImage from "../../../public/bg/home.jpg";

const mockedProps = {
  image: MockedImage,
  color: "yellow",
  title: "Title",
  description: "Description",
  buttonText: "Button",
};

jest.mock('next/router', () => require('next-router-mock'));

jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line @next/next/no-img-element
    return <img {...props} alt="image" />;
  },
}));

describe("Service card unit test", () => {
  it("should render correctly", () => {
    render(<ServiceCard {...mockedProps} />);

    expect(screen.getByTestId("card-image")).toBeVisible();
    expect(screen.getByTestId("card-title")).toBeVisible();
    expect(screen.getByTestId("card-description")).toBeVisible();
    expect(screen.getByTestId("card-btn")).toBeVisible();

    expect(screen.getByTestId("card-title")).toHaveTextContent("Title");
    expect(screen.getByTestId("card-description")).toHaveTextContent(
      "Description"
    );
    expect(screen.getByTestId("card-btn")).toHaveTextContent("Button");
  });

  it("should match snapshot", () => {
    const container = render(<ServiceCard {...mockedProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should render correct background color and button appearance", () => {
    const containerDefault = render(
      <ServiceCard {...mockedProps} color=''/>
    );
    expect(screen.getByTestId("card-btn").className).toContain(
      "bg-white text-gray-600"
    );
    containerDefault.unmount();

    const containerYellow = render(
      <ServiceCard {...mockedProps} color="yellow" />
    );

    expect(screen.getByTestId("card").className).toContain("bg-yellow");
    expect(screen.getByTestId("card-btn").className).toContain(
      "bg-yellow-light text-white"
    );
    containerYellow.unmount();

    const containerRed = render(<ServiceCard {...mockedProps} color="red" />);
    expect(screen.getByTestId("card").className).toContain("bg-red");
    expect(screen.getByTestId("card-btn").className).toContain(
      "bg-red-light text-white"
    );
    containerRed.unmount();

    const containerBlue = render(<ServiceCard {...mockedProps} color="blue" />);
    expect(screen.getByTestId("card").className).toContain("bg-blue-lotus");
    expect(screen.getByTestId("card-btn").className).toContain(
      "bg-blue opacity-90 text-white"
    );
    containerBlue.unmount();
  });

  it("should disable button when there is no link (button href)", () => {
    render(<ServiceCard {...mockedProps} color="yellow" />);
    expect(screen.getByTestId("card-btn")).toBeDisabled();
  });

  it("should not disable button when there is a link (button href)", () => {
    render(
      <ServiceCard {...mockedProps} color="yellow" buttonHref="/donate" />
    );
    expect(screen.getByTestId("card-btn")).not.toBeDisabled();
  });

  it("should call window.open() when the button is clicked and isOpenNewPage is true", () => {
    render(
      <ServiceCard
        {...mockedProps}
        color="yellow"
        buttonHref="https://google.com"
        isOpenNewPage
      />
    );
    fireEvent.click(screen.getByTestId("card-btn"));
    expect(window.open).toHaveBeenCalledWith("https://google.com");
  });

  it("should redirect to correct page when the button is clicked", () => {
    mockRouter.setCurrentUrl("/");
    render(
      <ServiceCard
        {...mockedProps}
        color="yellow"
        buttonHref="donate"
      />
    );
    fireEvent.click(screen.getByTestId("card-btn"));
    expect(singletonRouter.pathname).toEqual("/donate")
  });
});
