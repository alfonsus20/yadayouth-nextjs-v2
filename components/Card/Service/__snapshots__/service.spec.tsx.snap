// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service card unit test should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative px-6 pt-20 md:pt-16 pb-8 flex-[0_1_400px] rounded-3xl text-white bg-yellow"
        data-testid="card"
      >
        <div
          class="absolute -top-12 left-10 bg-white rounded-xl"
        >
          <img
            alt="image"
            data-testid="card-image"
            height="85"
            src="home.jpg"
            width="85"
          />
        </div>
        <h3
          class="font-bold text-xl mb-1"
          data-testid="card-title"
        >
          Title
        </h3>
        <p
          class="text-lg mb-5"
          data-testid="card-description"
        >
          Description
        </p>
        <button
          class="hover:opacity-90 transition-opacity disabled:opacity-75 disabled:cursor-not-allowed py-2 rounded-full bg-yellow-light text-white px-6  text-md"
          data-testid="card-btn"
          disabled=""
        >
          Button
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative px-6 pt-20 md:pt-16 pb-8 flex-[0_1_400px] rounded-3xl text-white bg-yellow"
      data-testid="card"
    >
      <div
        class="absolute -top-12 left-10 bg-white rounded-xl"
      >
        <img
          alt="image"
          data-testid="card-image"
          height="85"
          src="home.jpg"
          width="85"
        />
      </div>
      <h3
        class="font-bold text-xl mb-1"
        data-testid="card-title"
      >
        Title
      </h3>
      <p
        class="text-lg mb-5"
        data-testid="card-description"
      >
        Description
      </p>
      <button
        class="hover:opacity-90 transition-opacity disabled:opacity-75 disabled:cursor-not-allowed py-2 rounded-full bg-yellow-light text-white px-6  text-md"
        data-testid="card-btn"
        disabled=""
      >
        Button
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
